File: development_v2/typescript_version/src/core/checkpointSystem.ts
// Auto-save checkpoint system with change detection and recovery points

import { EventEmitter } from 'events';
import { getDatabase } from '../data/database';
import { getLogger } from '../utils/logger';
import {
  CheckpointData,
  CheckpointEvent,
  RecoveryOptions,
  ProjectData
} from '../types';

export class CheckpointManager extends EventEmitter {
  private db = getDatabase();
  private logger = getLogger();
  private autoSaveInterval: NodeJS.Timeout | null = null;
  private activeProjects = new Set<string>();
  private lastCheckpointTime = new Map<string, number>();
  private changeDetector = new DataChangeTracker();
  private isInitialized = false;

  constructor(
    private autoSaveMinutes: number = 5,
    private maxCheckpointsPerProject: number = 50
  ) {
    super();
    this.setupEventHandlers();
  }

  async initialize(): Promise<void> {
    if (this.isInitialized) return;

    await this.db.initialize();
    this.startAutoSaveTimer();
    this.isInitialized = true;

    this.logger.info('Checkpoint manager initialized', {
      autoSaveMinutes: this.autoSaveMinutes,
      maxCheckpointsPerProject: this.maxCheckpointsPerProject
    });
  }

  private setupEventHandlers(): void {
    // Handle process termination for emergency checkpoints
    process.on('SIGINT', () => this.emergencyCheckpointAll());
    process.on('SIGTERM', () => this.emergencyCheckpointAll());

    // Handle uncaught exceptions
    process.on('uncaughtException', () => this.emergencyCheckpointAll());
  }

  private startAutoSaveTimer(): void {
    this.autoSaveInterval = setInterval(() => {
      this.performAutoCheckpoints();
    }, this.autoSaveMinutes * 60 * 1000);
  }

  private async performAutoCheckpoints(): Promise<void> {
    const activeProjects = Array.from(this.activeProjects);

    for (const projectId of activeProjects) {
      try {
        const lastCheckpoint = this.lastCheckpointTime.get(projectId) || 0;
        const timeSinceLastCheckpoint = Date.now() - lastCheckpoint;

        if (timeSinceLastCheckpoint >= this.autoSaveMinutes * 60 * 1000) {
          await this.createCheckpoint(projectId, 'auto_save');
        }
      } catch (error) {
        this.logger.logError(error as Error, {
          operation: 'auto_checkpoint',
          projectId
        });
      }
    }
  }

  async createCheckpoint(
    projectId: string,
    operation: string,
    data?: Partial<ProjectData>
  ): Promise<string> {
    const checkpointId = `checkpoint_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;

    try {
      // Get current project data if not provided
      let checkpointData = data;
      if (!checkpointData) {
        const project = await this.db.loadProject(projectId);
        if (project) {
          checkpointData = project;
        } else {
          throw new Error(`Project ${projectId} not found for checkpoint`);
        }
      }

      const jsonData = JSON.stringify(checkpointData);
      const dataSize = Buffer.byteLength(jsonData, 'utf8');

      const checkpoint: CheckpointData = {
        id: checkpointId,
        projectId,
        data: checkpointData,
        dataSize,
        operation,
        timestamp: new Date()
      };

      await this.db.createCheckpoint(checkpoint);

      // Update tracking
      this.lastCheckpointTime.set(projectId, Date.now());

      // Cleanup old checkpoints
      await this.cleanupOldCheckpoints(projectId);

      // Emit event
      const event: CheckpointEvent = {
        trigger: 'manual_save',
        projectId,
        data: checkpointData
      };
      this.emit('checkpointCreated', event);

      this.logger.logCheckpoint(checkpointId, dataSize, operation);

      return checkpointId;

    } catch (error) {
      this.logger.logError(error as Error, {
        operation: 'create_checkpoint',
        projectId,
        checkpointId
      });
      throw error;
    }
  }

  private async cleanupOldCheckpoints(projectId: string): Promise<void> {
    try {
      const checkpoints = await this.db.getCheckpoints(projectId, this.maxCheckpointsPerProject + 10);

      if (checkpoints.length > this.maxCheckpointsPerProject) {
        const excessCount = checkpoints.length - this.maxCheckpointsPerProject;

        // In a full implementation, you would delete excess checkpoints
        // For now, just log the need for cleanup
        this.logger.logRecoveryAction(
          `Checkpoint cleanup needed: ${excessCount} excess checkpoints`,
          true,
          { projectId }
        );
      }
    } catch (error) {
      this.logger.logError(error as Error, {
        operation: 'cleanup_checkpoints',
        projectId
      });
    }
  }

  async restoreFromCheckpoint(checkpointId: string): Promise<ProjectData | null> {
    try {
      const checkpoint = await this.db.get(`SELECT * FROM checkpoints WHERE id = ?`, [checkpointId]);

      if (!checkpoint) {
        return null;
      }

      const checkpointData: CheckpointData = {
        id: checkpoint.id,
        projectId: checkpoint.project_id,
        data: JSON.parse(checkpoint.data),
        dataSize: checkpoint.data_size,
        operation: checkpoint.operation,
        timestamp: new Date(checkpoint.created_at)
      };

      this.logger.logRecoveryAction(
        `Restored from checkpoint ${checkpointId}`,
        true,
        { checkpointId, operation: checkpointData.operation }
      );

      return checkpointData.data as ProjectData;

    } catch (error) {
      this.logger.logError(error as Error, {
        operation: 'restore_from_checkpoint',
        checkpointId
      });
      return null;
    }
  }

  async getRecoveryOptions(projectId: string): Promise<RecoveryOptions[]> {
    const options: RecoveryOptions[] = [];

    try {
      // Get checkpoints
      const checkpoints = await this.db.getCheckpoints(projectId, 5);

      for (const checkpoint of checkpoints) {
        options.push({
          type: 'checkpoint',
          id: checkpoint.id,
          timestamp: checkpoint.timestamp,
          operation: checkpoint.operation,
          description: `Restore from ${checkpoint.operation || 'checkpoint'} (${checkpoint.timestamp.toLocaleString()})`,
          dataSize: checkpoint.dataSize
        });
      }

      // Add last good state option
      const project = await this.db.loadProject(projectId);
      if (project) {
        options.push({
          type: 'last_good_state',
          id: 'last_good_state',
          timestamp: project.updatedAt,
          description: 'Restore from last successfully saved state'
        });
      }

      return options;

    } catch (error) {
      this.logger.logError(error as Error, {
        operation: 'get_recovery_options',
        projectId
      });
      return [];
    }
  }

  async trackDataChange(projectId: string, newData: ProjectData, forceCheckpoint: boolean = false): Promise<void> {
    const hasSignificantChange = this.changeDetector.hasSignificantChange(projectId, newData);

    if (forceCheckpoint || hasSignificantChange) {
      await this.createCheckpoint(projectId, 'data_change', newData);
    }
  }

  markProjectActive(projectId: string): void {
    this.activeProjects.add(projectId);
    this.lastCheckpointTime.set(projectId, Date.now());
  }

  markProjectInactive(projectId: string): void {
    this.activeProjects.delete(projectId);
    this.lastCheckpointTime.delete(projectId);
  }

  async emergencyCheckpointAll(): Promise<void> {
    this.logger.logRecoveryAction('Emergency checkpoint initiated', true);

    const activeProjects = Array.from(this.activeProjects);

    for (const projectId of activeProjects) {
      try {
        await this.createCheckpoint(projectId, 'emergency_save');
      } catch (error) {
        this.logger.logError(error as Error, {
          operation: 'emergency_checkpoint',
          projectId
        });
      }
    }
  }

  getActiveProjects(): string[] {
    return Array.from(this.activeProjects);
  }

  async forceCheckpointAllActive(): Promise<void> {
    const activeProjects = Array.from(this.activeProjects);

    for (const projectId of activeProjects) {
      try {
        await this.createCheckpoint(projectId, 'forced_save');
      } catch (error) {
        this.logger.logError(error as Error, {
          operation: 'force_checkpoint',
          projectId
        });
      }
    }
  }

  shutdown(): void {
    if (this.autoSaveInterval) {
      clearInterval(this.autoSaveInterval);
      this.autoSaveInterval = null;
    }

    this.emergencyCheckpointAll();
    this.logger.info('Checkpoint manager shutdown');
  }
}

class DataChangeTracker {
  private lastHashes = new Map<string, string>();
  private changeThreshold: number;

  constructor(changeThreshold: number = 0.1) {
    this.changeThreshold = changeThreshold;
  }

  hasSignificantChange(projectId: string, newData: ProjectData): boolean {
    const currentHash = this.calculateDataHash(newData);
    const lastHash = this.lastHashes.get(projectId);

    if (!lastHash) {
      this.lastHashes.set(projectId, currentHash);
      return true; // First time seeing this project
    }

    const hasChanged = currentHash !== lastHash;

    if (hasChanged) {
      this.lastHashes.set(projectId, currentHash);
    }

    return hasChanged;
  }

  private calculateDataHash(data: ProjectData): string {
    // Simple hash for change detection
    const dataString = JSON.stringify(data, Object.keys(data).sort());
    let hash = 0;

    for (let i = 0; i < dataString.length; i++) {
      const char = dataString.charCodeAt(i);
      hash = ((hash << 5) - hash) + char;
      hash = hash & hash; // Convert to 32-bit integer
    }

    return hash.toString();
  }

  clearTracking(projectId: string): void {
    this.lastHashes.delete(projectId);
  }
}

// Decorators for checkpoint integration
export function checkpointOnChange(projectId?: string) {
  return function(target: any, propertyKey: string, descriptor: PropertyDescriptor) {
    const originalMethod = descriptor.value;
    const logger = getLogger();

    descriptor.value = async function(...args: any[]): Promise<any> {
      const result = await originalMethod.apply(this, args);

      // Try to create checkpoint if project data changed
      try {
        if (this.checkpointManager && this.getCurrentProjectData) {
          const currentData = this.getCurrentProjectData();
          const projId = projectId || this.currentProjectId;

          if (projId && currentData) {
            await this.checkpointManager.trackDataChange(projId, currentData);
          }
        }
      } catch (error) {
        logger.logError(error as Error, {
          operation: 'checkpoint_on_change_decorator',
          inputData: { method: propertyKey }
        });
      }

      return result;
    };

    return descriptor;
  };
}

// Global checkpoint manager instance
let checkpointManagerInstance: CheckpointManager | null = null;

export function getCheckpointManager(): CheckpointManager {
  if (!checkpointManagerInstance) {
    checkpointManagerInstance = new CheckpointManager();
  }
  return checkpointManagerInstance;
}

export function initializeCheckpointSystem(): Promise<void> {
  return getCheckpointManager().initialize();
}

// Convenience functions
export async function createCheckpoint(projectId: string, operation: string, data?: Partial<ProjectData>): Promise<string> {
  return getCheckpointManager().createCheckpoint(projectId, operation, data);
}

export async function getRecoveryOptions(projectId: string): Promise<RecoveryOptions[]> {
  return getCheckpointManager().getRecoveryOptions(projectId);
}

export function shutdownCheckpointSystem(): void {
  if (checkpointManagerInstance) {
    checkpointManagerInstance.shutdown();
  }
}

File: development_v2/valley_calculator/core/recovery/checkpoint_system.py
# checkpoint_system.py - Auto-save and checkpoint management
# Provides incremental saves and recovery points to prevent data loss

import time
import threading
import uuid
from typing import Dict, Any, Optional, List, Callable
from datetime import datetime, timedelta

from ...utils.logging.logger import get_logger
from ...data.persistence.database import get_database
from .error_handlers import resilient_operation, auto_save


class CheckpointManager:
    """
    Manages automatic checkpoints and data recovery points.

    Features:
    - Automatic periodic checkpoints
    - Change-based checkpoints (when significant changes occur)
    - Recovery point management and restoration
    - Performance-aware checkpointing
    """

    def __init__(self, auto_save_interval: int = 300, max_checkpoints: int = 50):
        """
        Initialize the checkpoint manager.

        Args:
            auto_save_interval: Seconds between automatic checkpoints
            max_checkpoints: Maximum number of checkpoints to keep per project
        """
        self.auto_save_interval = auto_save_interval
        self.max_checkpoints = max_checkpoints
        self.logger = get_logger()
        self.db = get_database()

        # Checkpoint state
        self.last_checkpoint_time = time.time()
        self.checkpoint_lock = threading.Lock()
        self.active_projects = {}  # project_id -> last_change_time

        # Auto-save thread
        self.auto_save_thread = None
        self.stop_auto_save = threading.Event()

        # Start auto-save daemon
        self._start_auto_save_daemon()

    def _start_auto_save_daemon(self):
        """Start the background auto-save daemon."""
        self.auto_save_thread = threading.Thread(
            target=self._auto_save_worker,
            daemon=True,
            name="CheckpointAutoSave"
        )
        self.auto_save_thread.start()

    def _auto_save_worker(self):
        """Background worker for automatic checkpoints."""
        while not self.stop_auto_save.wait(self.auto_save_interval):
            try:
                self._perform_auto_checkpoints()
            except Exception as e:
                self.logger.log_error(e, operation="auto_checkpoint_worker")

    def _perform_auto_checkpoints(self):
        """Perform automatic checkpoints for active projects."""
        current_time = time.time()

        with self.checkpoint_lock:
            # Check each active project
            for project_id, last_change in self.active_projects.items():
                if current_time - last_change > self.auto_save_interval:
                    try:
                        # Get current project data (this would be implemented by subclasses)
                        project_data = self._get_project_data(project_id)
                        if project_data:
                            self.create_checkpoint(
                                project_id,
                                project_data,
                                f"auto_save_{int(current_time)}",
                                "auto_save"
                            )
                    except Exception as e:
                        self.logger.log_error(e, operation="auto_checkpoint",
                                            context={"project_id": project_id})

    def create_checkpoint(self, project_id: str, data: Dict[str, Any],
                         checkpoint_id: str = None, operation: str = None) -> bool:
        """
        Create a checkpoint for the given data.

        Args:
            project_id: Project identifier
            data: Data to checkpoint
            checkpoint_id: Optional checkpoint identifier
            operation: Operation that triggered the checkpoint

        Returns:
            True if checkpoint was created successfully
        """
        if checkpoint_id is None:
            checkpoint_id = f"checkpoint_{int(time.time())}_{uuid.uuid4().hex[:8]}"

        try:
            success = self.db.create_checkpoint(project_id, checkpoint_id, data, operation)

            if success:
                # Update last checkpoint time
                with self.checkpoint_lock:
                    self.last_checkpoint_time = time.time()
                    self.active_projects[project_id] = time.time()

                # Cleanup old checkpoints
                self._cleanup_old_checkpoints(project_id)

                self.logger.log_checkpoint(checkpoint_id, len(str(data)), operation)

            return success

        except Exception as e:
            self.logger.log_error(e, operation="create_checkpoint",
                                context={"project_id": project_id, "checkpoint_id": checkpoint_id})
            return False

    def _cleanup_old_checkpoints(self, project_id: str):
        """Clean up old checkpoints to prevent storage bloat."""
        try:
            checkpoints = self.db.get_recent_checkpoints(project_id, self.max_checkpoints + 10)

            if len(checkpoints) > self.max_checkpoints:
                # This would need database schema changes to support deletion
                # For now, just log the need for cleanup
                excess_count = len(checkpoints) - self.max_checkpoints
                self.logger.log_recovery_action(
                    f"Checkpoint cleanup needed: {excess_count} excess checkpoints for {project_id}",
                    True
                )

        except Exception as e:
            self.logger.log_error(e, operation="cleanup_checkpoints",
                                context={"project_id": project_id})

    def restore_from_checkpoint(self, checkpoint_id: str) -> Optional[Dict[str, Any]]:
        """
        Restore data from a checkpoint.

        Args:
            checkpoint_id: Checkpoint identifier

        Returns:
            Restored data or None if checkpoint not found
        """
        try:
            data = self.db.restore_from_checkpoint(checkpoint_id)

            if data:
                self.logger.log_recovery_action(
                    f"Successfully restored from checkpoint {checkpoint_id}",
                    True
                )
            else:
                self.logger.log_recovery_action(
                    f"Failed to restore from checkpoint {checkpoint_id}",
                    False
                )

            return data

        except Exception as e:
            self.logger.log_error(e, operation="restore_checkpoint",
                                context={"checkpoint_id": checkpoint_id})
            return None

    def get_recovery_options(self, project_id: str) -> List[Dict[str, Any]]:
        """
        Get available recovery options for a project.

        Args:
            project_id: Project identifier

        Returns:
            List of recovery options with metadata
        """
        try:
            checkpoints = self.db.get_recent_checkpoints(project_id, 10)

            recovery_options = []
            for checkpoint in checkpoints:
                recovery_options.append({
                    'type': 'checkpoint',
                    'id': checkpoint['checkpoint_id'],
                    'timestamp': checkpoint['created_at'],
                    'operation': checkpoint['operation'],
                    'data_size': checkpoint['data_size']
                })

            # Add option to restore from last known good state
            recovery_options.append({
                'type': 'last_good_state',
                'description': 'Restore from last successfully saved state'
            })

            return recovery_options

        except Exception as e:
            self.logger.log_error(e, operation="get_recovery_options",
                                context={"project_id": project_id})
            return []

    def mark_project_active(self, project_id: str):
        """
        Mark a project as active for auto-checkpointing.

        Args:
            project_id: Project identifier
        """
        with self.checkpoint_lock:
            self.active_projects[project_id] = time.time()

    def mark_project_inactive(self, project_id: str):
        """
        Mark a project as inactive (removes from auto-checkpointing).

        Args:
            project_id: Project identifier
        """
        with self.checkpoint_lock:
            self.active_projects.pop(project_id, None)

    def force_checkpoint_all_active(self):
        """Force immediate checkpoint for all active projects."""
        with self.checkpoint_lock:
            active_projects = list(self.active_projects.keys())

        for project_id in active_projects:
            try:
                project_data = self._get_project_data(project_id)
                if project_data:
                    self.create_checkpoint(
                        project_id,
                        project_data,
                        f"forced_{int(time.time())}",
                        "forced_save"
                    )
            except Exception as e:
                self.logger.log_error(e, operation="force_checkpoint",
                                    context={"project_id": project_id})

    def _get_project_data(self, project_id: str) -> Optional[Dict[str, Any]]:
        """
        Get current project data. This should be overridden by subclasses.

        Args:
            project_id: Project identifier

        Returns:
            Current project data
        """
        # This is a placeholder - in a real implementation, this would
        # get the current state from the application
        return None

    def shutdown(self):
        """Shutdown the checkpoint system gracefully."""
        self.logger.log_recovery_action("Checkpoint system shutting down", True)

        # Stop auto-save thread
        self.stop_auto_save.set()
        if self.auto_save_thread and self.auto_save_thread.is_alive():
            self.auto_save_thread.join(timeout=5.0)

        # Force final checkpoints for all active projects
        self.force_checkpoint_all_active()


class DataChangeTracker:
    """
    Tracks data changes to determine when checkpoints are needed.

    This class monitors data modifications and triggers checkpoints
    when significant changes occur.
    """

    def __init__(self, checkpoint_manager: CheckpointManager,
                 change_threshold: float = 0.1):
        """
        Initialize the data change tracker.

        Args:
            checkpoint_manager: Checkpoint manager instance
            change_threshold: Threshold for triggering checkpoints (0.1 = 10% change)
        """
        self.checkpoint_manager = checkpoint_manager
        self.change_threshold = change_threshold
        self.last_data_hash = None
        self.change_lock = threading.Lock()

    def track_change(self, project_id: str, new_data: Dict[str, Any],
                    force_checkpoint: bool = False):
        """
        Track a data change and potentially trigger a checkpoint.

        Args:
            project_id: Project identifier
            new_data: New data state
            force_checkpoint: Force checkpoint regardless of change threshold
        """
        with self.change_lock:
            current_hash = self._calculate_data_hash(new_data)

            if force_checkpoint or self._has_significant_change(current_hash):
                self.checkpoint_manager.create_checkpoint(
                    project_id,
                    new_data,
                    operation="data_change"
                )
                self.last_data_hash = current_hash

    def _calculate_data_hash(self, data: Dict[str, Any]) -> str:
        """Calculate a simple hash of the data for change detection."""
        import hashlib
        import json

        # Create a normalized JSON string for consistent hashing
        normalized = json.dumps(data, sort_keys=True, default=str)
        return hashlib.md5(normalized.encode()).hexdigest()

    def _has_significant_change(self, new_hash: str) -> bool:
        """Determine if the change is significant enough for a checkpoint."""
        if self.last_data_hash is None:
            return True

        # Simple hash comparison - in a more sophisticated implementation,
        # you could calculate actual change percentage
        return new_hash != self.last_data_hash


# Decorators for checkpoint integration
def checkpoint_on_change(project_id_param: str = "project_id"):
    """
    Decorator that creates checkpoints when data changes significantly.

    Args:
        project_id_param: Parameter name containing the project ID
    """
    def decorator(func: Callable[..., Any]) -> Callable[..., Any]:
        @functools.wraps(func)
        def wrapper(*args, **kwargs) -> Any:
            # Get project ID from parameters
            project_id = None
            if project_id_param in kwargs:
                project_id = kwargs[project_id_param]
            else:
                # Try to find it in positional args by inspecting function signature
                import inspect
                sig = inspect.signature(func)
                param_names = list(sig.parameters.keys())
                if project_id_param in param_names:
                    idx = param_names.index(project_id_param)
                    if idx < len(args):
                        project_id = args[idx]

            result = func(*args, **kwargs)

            # Create checkpoint if we have project_id and result indicates data change
            if project_id and result is not None:
                try:
                    checkpoint_mgr = get_checkpoint_manager()
                    if hasattr(checkpoint_mgr, 'data_tracker'):
                        checkpoint_mgr.data_tracker.track_change(project_id, result)
                except Exception as e:
                    logger.log_error(e, operation="checkpoint_on_change")

            return result

        return wrapper
    return decorator


# Global checkpoint manager instance
_checkpoint_manager_instance = None
_checkpoint_manager_lock = threading.Lock()


def get_checkpoint_manager() -> CheckpointManager:
    """Get the global checkpoint manager instance."""
    global _checkpoint_manager_instance
    with _checkpoint_manager_lock:
        if _checkpoint_manager_instance is None:
            _checkpoint_manager_instance = CheckpointManager()
        return _checkpoint_manager_instance


# Convenience functions
def create_checkpoint(project_id: str, data: Dict[str, Any], **kwargs) -> bool:
    """Create a checkpoint for the given data."""
    return get_checkpoint_manager().create_checkpoint(project_id, data, **kwargs)


def get_recovery_options(project_id: str) -> List[Dict[str, Any]]:
    """Get recovery options for a project."""
    return get_checkpoint_manager().get_recovery_options(project_id)
